segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java

```diff
@@ -552,6 +571,144 @@ public void updateSegmentPolicy(UpdateSegmentPolicy updateSegmentPolicy) {
+                                                                         e.getKey().getKeyVersion()))
+                                                     .collect(Collectors.toList());
+        tableStore.put(segment, entries, TIMEOUT)
+                  .thenAccept(versions -> connection.send(new WireCommands.TableEntriesUpdated(updateTableEntries.getRequestId(), versions)))
+                  .exceptionally(e -> handleException(updateTableEntries.getRequestId(), segment, operation, e));
+    }
+
+    @Override
+    public void removeTableKeys(final WireCommands.RemoveTableKeys removeTableKeys) {
+        String segment = removeTableKeys.getSegment();
+        final String operation = "removeTableKeys";
+
+        if (!verifyToken(segment, removeTableKeys.getRequestId(), removeTableKeys.getDelegationToken(), operation)) {
+            return;
+        }
+
+        log.info(removeTableKeys.getRequestId(), "Removing table keys {}.", removeTableKeys);
+
+        List<TableKey> keys = removeTableKeys.getKeys().stream()
+                                                .map(k -> TableKey.versioned(new ByteArraySegment(k.getData()), k.getKeyVersion()))
+                                                .collect(Collectors.toList());
+        tableStore.remove(segment, keys, TIMEOUT)
+                  .thenRun(() -> connection.send(new WireCommands.TableKeysRemoved(removeTableKeys.getRequestId(), segment)))
+                  .exceptionally(e -> handleException(removeTableKeys.getRequestId(), segment, operation, e));
+    }
+
+
+    @Override
+    public void readTable(final WireCommands.ReadTable readTable) {
+        final String segment = readTable.getSegment();
+        final String operation = "readTable";
+
+        if (!verifyToken(segment, readTable.getRequestId(), readTable.getDelegationToken(), operation)) {
+            return;
+        }
+
+        log.info(readTable.getRequestId(), "Reading from table {}.", readTable);
+
+        final List<ArrayView> keys = readTable.getKeys().stream()
+                                              .map(k -> new ByteArraySegment(k.getData().array()))
+                                              .collect(Collectors.toList());
+        tableStore.get(segment, keys, TIMEOUT)
+                  .thenAccept(values -> connection.send(new WireCommands.TableRead(readTable.getRequestId(), segment, getTableEntries(values))))
+                  .exceptionally(e -> handleException(readTable.getRequestId(), segment, operation, e));
+    }
+
+    private WireCommands.TableEntries getTableEntries(final List<TableEntry> updateData) {
+
+        final List<Map.Entry<WireCommands.TableKey, WireCommands.TableValue>> entries = updateData.stream().map(te -> {
+            if (te == null) {
+                // no entry for the corresponding key.
+                return new AbstractMap.SimpleImmutableEntry<>(WireCommands.TableKey.EMPTY, WireCommands.TableValue.EMPTY);
```

Comment:
TODO: the key  should correspond to the key passed in the request.

Reply:
Yes, this has to be done.

Reply:
done.